src/ReachSets/discretize.jl

```diff
@@ -1,310 +1,656 @@
+uses less conservative bounds.
+
+[1] Le Guernic, C., & Girard, A., 2010, *Reachability analysis of linear systems
+using support functions. Nonlinear Analysis: Hybrid Systems, 4(2), 250-262.*
+"""
+function discretize_firstorder(𝑆::InitialValueProblem,
+                                δ::Float64;
+                                p::Float64=Inf,
+                                exp_method::String="base")
+
+    # unwrap coefficient matrix and initial states
+    A, X0 = 𝑆.s.A, 𝑆.x0 
+
+    # system size; A is assumed square
+    n = size(A, 1)
+
+    Anorm = norm(Matrix(A), p)
+    κ = exp(δ*Anorm) - 1.0 - δ*Anorm
+    RX0 = norm(X0, p)
+
+    # compute exp(A*δ)
+    ϕ = exp_Aδ(A, δ, exp_method=exp_method)
+
+    if inputdim(𝑆) == 0
+        α = κ * RX0
+        □α = Ballp(p, zeros(n), α)
+        Ω0 = ConvexHull(X0, ϕ * X0 ⊕ □α)
+        return IVP(LDS(ϕ), Ω0)
+
+    elseif isaffine(𝑆)
+        Uset = inputset(𝑆)
         if Uset isa ConstantInput
             U = next_set(Uset)
-            RU = norm(U, Inf)
-            α = (exp(δ*Anorm) - 1. - δ*Anorm)*(RX0 + RU/Anorm)
-            β = (exp(δ*Anorm) - 1. - δ*Anorm)*RU/Anorm
-            Ω0 = CH(X0, ϕ * X0 + δ * U + Ball2(zeros(size(ϕ, 1)), α))
-            discr_U =  δ * U + Ball2(zeros(size(ϕ, 1)), β)
-            return DiscreteSystem(ϕ, Ω0, discr_U)
+            RU = norm(U, p)
+
+            # bloating coefficients
+            α = κ*(RX0 + RU/Anorm)
+            β = κ*RU/Anorm
+
+            # transformation of the initial states
+            □α = Ballp(p, zeros(n), α)
+            Ω0 = ConvexHull(X0, ϕ*X0 ⊕ δ*U ⊕ □α)
+
+            # transformation of the inputs
+            □β = Ballp(p, zeros(n), β)
+            Ud = map(u -> δ*u ⊕ □β, Uset)
+            return IVP(CLCDS(ϕ, Id(n), nothing, Ud), Ω0)
+
         elseif Uset isa VaryingInput
-            discr_U = Vector{LazySet}(undef, length(Uset))
+            Ud = Vector{LazySet}(undef, length(Uset))
             for (i, Ui) in enumerate(Uset)
-                RU = norm(Ui, Inf)
-                α = (exp(δ*Anorm) - 1. - δ*Anorm)*(RX0 + RU/Anorm)
-                β = (exp(δ*Anorm) - 1. - δ*Anorm)*RU/Anorm
-                Ω0 = CH(X0, ϕ * X0 + δ * Ui + Ball2(zeros(size(ϕ, 1)), α))
-                discr_U[i] =  δ * Ui + Ball2(zeros(size(ϕ, 1)), β)
+                RU = norm(Ui, p)
+
+                # bloating factors
+                α = κ*(RX0 + RU/Anorm)
+                β = κ*RU/Anorm
+
+                if i == 1
+                    # transform initial states
+                    □α = Ballp(p, zeros(n), α)
+                    Ω0 = ConvexHull(X0, ϕ * X0 ⊕ δ * Ui ⊕ □α)
+                end
+
+                # transform inputs
+                □β = Ballp(p, zeros(n), β)
+                Ud[i] = δ*Ui ⊕ □β
             end
-            return DiscreteSystem(ϕ, Ω0, discr_U)
+            Ud = VaryingInput(Ud)
+            return IVP(CLCDS(ϕ, Id(n), nothing, Ud), Ω0)
         end
```

Comment:
Add an `else` case here. Currently, if the inputs are neither `ConstantInput` nor `VaryingInput`, the function returns something unexpected. (You can copy the message from the very bottom of the file.)

Reply:
These are the only two types of inputs, but yes, better to prevent from future changes. I've added this message which was already used in the last function :)